src/mrpro/algorithms/optimizers/pgd.py

```diff
@@ -0,0 +1,143 @@
+"""Proximal Gradient Descent algorithm."""
+
+import math
+
+import torch
+from typing_extensions import Unpack
+
+from mrpro.operators import ProximableFunctionalSeparableSum
+from mrpro.operators.Functional import ProximableFunctional
+from mrpro.operators.Operator import Operator
+
+
+def pgd(
+    f: Operator[torch.Tensor, tuple[torch.Tensor]] | Operator[Unpack[tuple[torch.Tensor, ...]], tuple[torch.Tensor]],
+    g: ProximableFunctional | ProximableFunctionalSeparableSum,
+    initial_value: torch.Tensor | tuple[torch.Tensor, ...],
+    stepsize: float = 1.0,
+    reg_parameter: float = 0.01,
+    max_iterations: int = 128,
+    backtrack_factor: float = 1.0,
+) -> tuple[torch.Tensor, ...]:
+    r"""Proximal gradient descent algorithm for solving problem min_x f(x) + g(x).
+
+    f is convex, differentiable, and with L-Lispchitz gradient.
+    g is convex, non-smooth with computable proximal map.
+
+    For fixed stepsize t, pgd converges globally when t \in (0, 1/L),
+    where L is the Lipschitz constant of the gradient of f.
+    In applications, f is usually of the form f(x) = 1/2 ||Ax - target||^2, where A is a linear operator.
+    In this case, t \in (0, 1/||A^T A||) for convergence.
+    If no backtracking is used, the fixed stepsize should be given accordingly to the convergence condition.
+
+    Example:
+        L1 regularized image reconstruction. Problem formulation: min_x 1/2 ||Fx - target||^2 + ||x||_1,
+        with F being the Fourier Transform, target the acquired data \in k-space and x \in image space,
+        f(x) = 1/2 ||Fx - target||^2, g(x) = ||x||_1.
+        In this case, ||F^T F|| = 1. ::
+                fft = FastFourierOp()
+                l2 = L2NormSquared(target=kspace_data)
+                f = l2 @ fft
+                g = L1Norm()
+                fft_norm = 1.
+                stepsize = 0.85 * 1 / fft_norm
+                initial_value = torch.ones(image_space_shape)
+                pgd_image_solution = pgd(
+                    f=f,
+                    g=g,
+                    initial_value=initial_value,
+                    stepsize=stepsize,
+                    reg_parameter=0.01,
+                    max_iterations=200,
+                    backtrack_factor=1.0,
+                )
+
+
+    Parameters
+    ----------
+    f
+        convex, differentiable functional
+    g
+        convex, non-smooth functional with computable proximal map
+    initial_value
+        initial value for the solution of the algorithm
+    stepsize
+        stepsize needed in the gradient step, is constant throughout all
```

Comment:
is it`doesnt the algrithm backtrack?

Reply:
Yes I'm stupid, I meant when no backtracking, I'll correct it now